gustaf/spline/microstructure/generator.py

```diff
@@ -0,0 +1,553 @@
+
+        In order to use said function, the Microtile needs to provide a couple
+        of attributes
+          1. `evaluation_points` - a list of points defined in the unit cube
+             that will be evaluated in the parametrization function to provide
+             the required set of data points
+          2. `parameter_space_dimension` - dimensionality of the
+              parametrization function and number of design variables for said
+              microtile
+
+        Parameters
+        ----------
+        parametrization_function: Callable
+
+        Returns
+        --------
+        None
+        """
+        if not callable(parametrization_function):
+            raise TypeError("parametrization_function must be callable")
+        self._parametrization_function = parametrization_function
+        self._sanity_check()
+
+    def create(self, closing_face=None, knot_span_wise=True, **kwargs):
+        """
+        Create a Microstructure
+
+        Parameters
+        ----------
+        closing_face : string
+          If not None, Microtile must provide a function `closing_tile`
+          Represents coordinate to be a closed surface {"x", "y", "z"}
+        knot_span_wise : bool
+          Insertion per knotspan vs. total number per paradim
+        **kwargs
+          will be passed to `create_tile` function
+
+        Returns
+        -------
+        Microstructure : list<spline>
+          finished microstructure based on object requirements
+        """
+        import itertools
+
+        # Check if all information is gathered
+        if not self._sanity_check():
+            raise ValueError("Not enough information provided, abort")
+        # check if user wants closed structure
+        closing_face_dim = {"x": 0, "y": 1, "z": 2}.get(closing_face)
+        is_closed = closing_face_dim is not None
+        if not is_closed and (closing_face is not None):
+            raise ValueError(
+                    "Invalid format for closing_face argument, (handed: "
+                    f"{closing_face}), must be one of"
+                    "{'x', 'y', 'z'}"
+            )
+
+        if is_closed:
+            is_closed = True
+            if closing_face_dim >= self._deformation_function.para_dim:
+                raise ValueError(
+                        "closing face must be smaller than the deformation "
+                        "function's parametric dimension"
+                )
+            if self._parametrization_function is None:
+                raise ValueError(
+                        "Faceclosure is currently only implemented for "
+                        "parametrized microstructures"
+                )
+
+        # Prepare the deformation function
+        # Transform into a non-uniform splinetype and make sure to work on copy
+        if hasattr(self._deformation_function, "bspline"):
+            deformation_function_copy_ = self._deformation_function.bspline
```

Comment:
```suggestion
            deformation_function_copy = self._deformation_function.bspline
```
what's the purpose of trailing underscore? It is usually just to avoid overlapping built-in variables